README.md
Chronic
=======

This is a fork of https://github.com/mojombo/chronic since that project
does not appear to be maintained.

Chronic is a natural language date/time parser written in pure Ruby. See below
for the wide variety of formats Chronic will parse.

## Installation

```
$ gem install chronic
```

## Usage

```ruby
require 'chronic'

Time.now   #=> Sun Aug 27 23:18:25 PDT 2006

Chronic.parse('tomorrow')
  #=> Mon Aug 28 12:00:00 PDT 2006

Chronic.parse('monday', :context => :past)
  #=> Mon Aug 21 12:00:00 PDT 2006

Chronic.parse('this tuesday 5:00')
  #=> Tue Aug 29 17:00:00 PDT 2006

Chronic.parse('this tuesday 5:00', :ambiguous_time_range => :none)
  #=> Tue Aug 29 05:00:00 PDT 2006

Chronic.parse('may 27th', :now => Time.local(2000, 1, 1))
  #=> Sat May 27 12:00:00 PDT 2000

Chronic.parse('may 27th', :guess => false)
  #=> Sun May 27 00:00:00 PDT 2007..Mon May 28 00:00:00 PDT 2007

Chronic.parse('6/4/2012', :endian_precedence => :little)
  #=> Fri Apr 06 00:00:00 PDT 2012

Chronic.parse('INVALID DATE')
  #=> nil
```

If the parser can find a date or time, either a Time or Chronic::Span
will be returned (depending on the value of `:guess`). If no
date or time can be found, `nil` will be returned.

See `Chronic.parse` for detailed usage instructions.

## Examples

Chronic can parse a huge variety of date and time formats. Following is a
small sample of strings that will be properly parsed. Parsing is case
insensitive and will handle common abbreviations and misspellings.

#### Simple

* thursday
* november
* summer
* friday 13:00
* mon 2:35
* 4pm
* 10 to 8
* 10 past 2
* half past 2
* 6 in the morning
* friday 1pm
* sat 7 in the evening
* yesterday
* today
* tomorrow
* last week
* next week
* this tuesday
* next month
* last winter
* this morning
* last night
* this second
* yesterday at 4:00
* last friday at 20:00
* last week tuesday
* tomorrow at 6:45pm
* afternoon yesterday
* thursday last week

#### Complex

* 3 years ago
* a year ago
* 5 months before now
* 7 hours ago
* 7 days from now
* 1 week hence
* in 3 hours
* 1 year ago tomorrow
* 3 months ago saturday at 5:00 pm
* 7 hours before tomorrow at noon
* 3rd wednesday in november
* 3rd month next year
* 3rd thursday this september
* 4th day last week
* fourteenth of june 2010 at eleven o'clock in the evening
* may seventh '97 at three in the morning

#### Specific Dates

* January 5
* 22nd of june
* 5th may 2017
* February twenty first
* dec 25
* may 27th
* October 2006
* oct 06
* jan 3 2010
* february 14, 2004
* february 14th, 2004
* 3 jan 2000
* 17 april 85
* 5/27/1979
* 27/5/1979
* 05/06
* 1979-05-27
* Friday
* 5
* 4:00
* 17:00
* 0800

#### Specific Times (many of the above with an added time)

* January 5 at 7pm
* 22nd of june at 8am
* 1979-05-27 05:00:00
* 03/01/2012 07:25:09.234567
* 2013-08-01T19:30:00.345-07:00
* 2013-08-01T19:30:00.34-07:00
* etc


## Time Zones

Chronic allows you to set which Time class to use when constructing times. By
default, the built in Ruby time class creates times in your system's local
time zone. You can set this to something like ActiveSupport's
[TimeZone](http://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html)
class to get full time zone support.

```
>> Time.zone = "UTC"
>> Chronic.time_class = Time.zone
>> Chronic.parse("June 15 2006 at 5:45 AM")
=> Thu, 15 Jun 2006 05:45:00 UTC +00:00
```

## Limitations

Chronic uses Ruby's built in Time class for all time storage and computation.
Because of this, only times that the Time class can handle will be properly
parsed. Parsing for times outside of this range will simply return `nil`.
Support for a wider range of times is planned for a future release.


## Contribute

If you'd like to hack on Chronic, start by forking the repo on GitHub:

https://github.com/mojombo/chronic

The best way to get your changes merged back into core is as follows:

1. Clone down your fork
1. Create a thoughtfully named topic branch to contain your change
1. Install the development dependencies by running `bundle install`
1. Hack away
1. Add tests and make sure everything still passes by running `bundle exec rake`
1. Ensure your tests pass in multiple timezones. ie `TZ=utc bundle exec rake` `TZ=BST bundle exec rake`
1. If you are adding new functionality, document it in the README
1. Do not change the version number, we will do that on our end
1. If necessary, rebase your commits into logical chunks, without errors
1. Push the branch up to GitHub
1. Send a pull request for your branch


HISTORY.md
## --- Begin chronic changes ---

# 0.10.5 / 2020-02-26
* Rename gitlab-chronic -> chronic


# 0.10.4 / 2019-11-24

* Rename chronic -> gitlab-chronic (https://gitlab.com/gitlab-org/gitlab-chronic/merge_requests/5)

# 0.10.3 / 2019-11-23

* Fix Daylight Savings parsing (https://gitlab.com/gitlab-org/gitlab-chronic/merge_requests/3)
* Add .gitlab-ci.yml (https://gitlab.com/gitlab-org/gitlab-chronic/merge_requests/1)

## --- End gitlab-chronic changes ---

# 0.10.2 / 2013-09-09

* Fix 1.8.7 support (due to be dropped in 0.11.0)
* Bugfix for times with negative zones

# 0.10.1 / 2013-08-27

* Support `ActiveSupport::TimeZone` (#209, #208)

# 0.10.0 / 2013-08-25

* Chronic will parse subseconds correctly
  for all supported date/time formats (#195, #198 and #200)
* Support for date format: dd.mm.yyyy (#197)
* Option `:hours24` to parse as 24 hour clock (#201 and #202)
* `:guess` option allows to specify which part of Span to return.
  (accepted values `false`,`true`,`:begin`, `:middle`, `:end`)
* Replace `rcov` with `SimpleCov` for coverage generation
* Add more tests
* Various changes in codebase (#202 and #206)

# 0.9.1 / 2013-02-25

* Ensure Chronic strips periods from day portions (#173)
* Properly numerize "twelfth", "twentieth" etc. (#172, James McKinney)
* Ensure Chronic is compatible with Ruby 2.0.0 (#165, Ravil Bayramgalin)

# 0.9.0 / 2012-12-21

* Implement Chronic::Parser class and create an instance of this class
  instead of leaving all data in the class level of Chronic
* Various bug fixes
* Add support for excel date formats (#149, @jmondo)
* Added support for time expressions such as '10 till' or 'half
  past two' (#146, @chicagogrooves)
* Add support for RepeaterDayName, RepeaterMonthName,
  Ordinal/ScalarDay and Time (#153, @kareemk)

# 0.8.0 / 2012-09-16

* Support parsing "<ordinal> of this month" (#109)
* Support parsing ISO 8601 format (#115)
* Support parsing "on <day>" without a timestamp (#117)
* Fix time parsing regexp (#125)
* Support time when parsing dd-mm-yyy <time> (#126)
* Allow anchor handler to accept any separators (at, on) (#128)
* Support parsing EXIF date format (#112)
* Start using minitest for testing
* Ensure periods are interpreted as colons (#81).
* Support month/day and day/month parsing (#59).
* Support day(scalar)-month(name)-year(scalar) (#99).
* Handle text starting with 'a' or 'an' (#101, @steveburkett).
* Ensure post medium timestamps are correctly formatted (#89)

# 0.6.7 / 2012-01-31

* Handle day, month names with scalar day and year (Joe Fiorini)
* Ensure 31st parses correctly with day names (Joe Fiorini)

# 0.6.6 / 2011-11-23

* `Chronic.parse('thur')` no longer returns `nil` (@harold)

# 0.6.5 / 2011-11-04

* Fix bug when parsing ordinal repeaters (#73)
* Added handler support for day_name month_name (@imme5150)
* Fix bug when parsing strings prefixed with PM

# 0.6.4 / 2011-08-08

* Fixed bug where 'noon' was parsed as 00:00 rather than 12:00
  with :ambiguous_time_range => :none (Vladimir Chernis)
* Add support for handling '2009 May 22nd'
* Add the ability to handle scalar-day/repeater-month-name as well as ordinals

# 0.6.3 / 2011-08-01

* Ensure 'thu' is parsed as Thursday for 1.8.7 generic timestamp

# 0.6.2 / 2011-07-28

* Ensure specific endian handlers are prioritised over normal date handlers
* Recognize UTC as timezone and accept HH::MM timezone offset (Jason Dusek)

# 0.6.1 / 2011-07-21

* Ensure Handler definitions are executed in the correct order

# 0.6.0 / 2011-07-19

* Attempting to parse strings with days past the last day of a month will
  now return nil. ex: `Chronic.parse("30th February") #=> nil`
* All deprecated methods are marked for removal in Chronic 0.7.0
* Deprecated `Chronic.numericize_numbers` instead use
  `Chronic::Numerizer.numerize`
* Deprecated `Chronic::InvalidArgumentException` and instead use
  `ArgumentError`
* Deprecated `Time.construct` and use `Chronic.construct` in place of this
* Deprecated `Time#to_minidate`, instead use `Chronic::MiniDate.from_time(time)`
* Add support for handling generic timestamp for Ruby 1.9+

# 0.5.0 / 2011-07-01

* Replace commas with spaces instead of removing the char (Thomas Walpole)
* Added tests for RepeaterSeason
* Re-factored tests. Now rather than having a test_parsing method for testing
  all handlers, break them down independent of handler method. For example
  with handler `handle_sm_sd_sy` the subsequent test would be
  `test_handle_sm_sd_sy`
* Added support for parsing ordinal-dates/month-names/year, ie:
  `2nd of May 1995`
* Added support for parsing ordinal-dates and month names, ie:
  `22nd of February at 6:30pm`
* Fix `Time.construct` leap year checking. Instead use `Date.leap?(year)`

# 0.4.4 / 2011-06-12

* Fix RepeaterYear for fetching past year offsets when the current day is
  later than the last day of the same month in a past year (leap years) ie
  on 29th/feb (leap year) `last year` should (and now does) return 28th/feb
  instead of 1st/march
* Opt in for gem testing http://test.rubygems.org/

# 0.4.3 / 2011-06-08

* Fix issue with parsing 1:xxPM -- Ensure 1 is treated as ambiguous, not
  just >1

# 0.4.2 / 2011-06-07

* Fix MonthRepeater for fetching past month offsets when current day is
  later than the last day of a past month (ie on 29th of March when parsing
  `last month` Chronic would return March instead of February. Now Chronic
  returns the last day of the past month)

# 0.4.1 / 2011-06-05

* Fix MiniDate ranges for parsing seasons (Thomas Walpole)

# 0.4.0 / 2011-06-04

* Ensure context is being passed through grabbers. Now "Sunday at 2:18pm"
  with `:context => :past` will return the correct date
* Support parsing ordinal strings (eg first, twenty third => 1st, 23rd)
* Seasons now ignore DST and return 00 as an hour
* Support parsing 2 digit years and added `ambiguous_year_future_bias` option
* Support parsing 'thurs' for Thursday
* Fix pre_normalize() to remove periods before numerizing

lib/chronic/parser.rb
require 'chronic/dictionary'
require 'chronic/handlers'

module Chronic
  class Parser
    include Handlers

    # Hash of default configuration options.
    DEFAULT_OPTIONS = {
      :context => :future,
      :now => nil,
      :hours24 => nil,
      :week_start => :sunday,
      :guess => true,
      :ambiguous_time_range => 6,
      :endian_precedence    => [:middle, :little],
      :ambiguous_year_future_bias => 50
    }

    attr_accessor :now
    attr_reader :options

    # options - An optional Hash of configuration options:
    #        :context - If your string represents a birthday, you can set
    #                   this value to :past and if an ambiguous string is
    #                   given, it will assume it is in the past.
    #        :now - Time, all computations will be based off of time
    #               instead of Time.now.
    #        :hours24 - Time will be parsed as it would be 24 hour clock.
    #        :week_start - By default, the parser assesses weeks start on
    #                  sunday but you can change this value to :monday if
    #                  needed.
    #        :guess - By default the parser will guess a single point in time
    #                 for the given date or time. If you'd rather have the
    #                 entire time span returned, set this to false
    #                 and a Chronic::Span will be returned. Setting :guess to :end
    #                 will return last time from Span, to :middle for middle (same as just true)
    #                 and :begin for first time from span.
    #        :ambiguous_time_range - If an Integer is given, ambiguous times
    #                  (like 5:00) will be assumed to be within the range of
    #                  that time in the AM to that time in the PM. For
    #                  example, if you set it to `7`, then the parser will
    #                  look for the time between 7am and 7pm. In the case of
    #                  5:00, it would assume that means 5:00pm. If `:none`
    #                  is given, no assumption will be made, and the first
    #                  matching instance of that time will be used.
    #        :endian_precedence - By default, Chronic will parse "03/04/2011"
    #                 as the fourth day of the third month. Alternatively you
    #                 can tell Chronic to parse this as the third day of the
    #                 fourth month by setting this to [:little, :middle].
    #        :ambiguous_year_future_bias - When parsing two digit years
    #                 (ie 79) unlike Rubys Time class, Chronic will attempt
    #                 to assume the full year using this figure. Chronic will
    #                 look x amount of years into the future and past. If the
    #                 two digit year is `now + x years` it's assumed to be the
    #                 future, `now - x years` is assumed to be the past.
    def initialize(options = {})
      validate_options!(options)
      @options = DEFAULT_OPTIONS.merge(options)
      @now = options[:now] || Chronic.time_class.now
    end

    # Parse "text" with the given options
    # Returns either a Time or Chronic::Span, depending on the value of options[:guess]
    def parse(text)
      tokens = tokenize(text, options)
      span = tokens_to_span(tokens, options.merge(:text => text))

      puts "+#{'-' * 51}\n| #{tokens}\n+#{'-' * 51}" if Chronic.debug

      guess(span, options[:guess]) if span
    end

    # Clean up the specified text ready for parsing.
    #
    # Clean up the string by stripping unwanted characters, converting
    # idioms to their canonical form, converting number words to numbers
    # (three => 3), and converting ordinal words to numeric
    # ordinals (third => 3rd)
    #
    # text - The String text to normalize.
    #
    # Examples:
    #
    #   Chronic.pre_normalize('first day in May')
    #     #=> "1st day in may"
    #
    #   Chronic.pre_normalize('tomorrow after noon')
    #     #=> "next day future 12:00"
    #
    #   Chronic.pre_normalize('one hundred and thirty six days from now')
    #     #=> "136 days future this second"
    #
    # Returns a new String ready for Chronic to parse.
    def pre_normalize(text)
      text = text.to_s.downcase
      text.gsub!(/\b(\d{1,2})\.(\d{1,2})\.(\d{4})\b/, '\3 / \2 / \1')
      text.gsub!(/\b([ap])\.m\.?/, '\1m')
      text.gsub!(/(\s+|:\d{2}|:\d{2}\.\d+)\-(\d{2}:?\d{2})\b/, '\1tzminus\2')
      text.gsub!(/\./, ':')
      text.gsub!(/([ap]):m:?/, '\1m')
      text.gsub!(/'(\d{2})\b/) do
        number = $1.to_i

        if Chronic::Date::could_be_year?(number)
          Chronic::Date::make_year(number, options[:ambiguous_year_future_bias])
        else
          number
        end
      end
      text.gsub!(/['"]/, '')
      text.gsub!(/,/, ' ')
      text.gsub!(/^second /, '2nd ')
      text.gsub!(/\bsecond (of|day|month|hour|minute|second|quarter)\b/, '2nd \1')
      text.gsub!(/\bthird quarter\b/, '3rd q')
      text.gsub!(/\bfourth quarter\b/, '4th q')
      text.gsub!(/quarters?(\s+|$)(?!to|till|past|after|before)/, 'q\1')
      text = Numerizer.numerize(text)
      text.gsub!(/\b(\d)(?:st|nd|rd|th)\s+q\b/, 'q\1')
      text.gsub!(/([\/\-\,\@])/) { ' ' + $1 + ' ' }
      text.gsub!(/(?:^|\s)0(\d+:\d+\s*pm?\b)/, ' \1')
      text.gsub!(/\btoday\b/, 'this day')
      text.gsub!(/\btomm?orr?ow\b/, 'next day')
      text.gsub!(/\byesterday\b/, 'last day')
      text.gsub!(/\bnoon|midday\b/, '12:00pm')
      text.gsub!(/\bmidnight\b/, '24:00')
      text.gsub!(/\bnow\b/, 'this second')
      text.gsub!('quarter', '15')
      text.gsub!('half', '30')
      text.gsub!(/(\d{1,2}) (to|till|prior to|before)\b/, '\1 minutes past')
      text.gsub!(/(\d{1,2}) (after|past)\b/, '\1 minutes future')
      text.gsub!(/\b(?:ago|before(?: now)?)\b/, 'past')
      text.gsub!(/\bthis (?:last|past)\b/, 'last')
      text.gsub!(/\b(?:in|during) the (morning)\b/, '\1')
      text.gsub!(/\b(?:in the|during the|at) (afternoon|evening|night)\b/, '\1')
      text.gsub!(/\btonight\b/, 'this night')
      text.gsub!(/\b\d+:?\d*[ap]\b/,'\0m')
      text.gsub!(/\b(\d{2})(\d{2})(am|pm)\b/, '\1:\2\3')
      text.gsub!(/(\d)([ap]m|oclock)\b/, '\1 \2')
      text.gsub!(/\b(hence|after|from)\b/, 'future')
      text.gsub!(/^\s?an? /i, '1 ')
      text.gsub!(/\b(\d{4}):(\d{2}):(\d{2})\b/, '\1 / \2 / \3') # DTOriginal
      text.gsub!(/\b0(\d+):(\d{2}):(\d{2}) ([ap]m)\b/, '\1:\2:\3 \4')
      text
    end

    # Guess a specific time within the given span.
    #
    # span - The Chronic::Span object to calcuate a guess from.
    #
    # Returns a new Time object.
    def guess(span, mode = :middle)
      return span if not mode
      return span.begin + span.width / 2 if span.width > 1 and (mode == true or mode == :middle)
      return span.end if mode == :end
      span.begin
    end

    # List of Handler definitions. See Chronic.parse for a list of options this
    # method accepts.
    #
    # options - An optional Hash of configuration options.
    #
    # Returns a Hash of Handler definitions.
    def definitions(options = {})
      SpanDictionary.new(options).definitions
    end

    private


    def validate_options!(options)
      given = options.keys.map(&:to_s).sort
      allowed = DEFAULT_OPTIONS.keys.map(&:to_s).sort
      allowed << "zone"
puts "************************"
puts "zone: #{options[:zone]}"
puts "************************"
      non_permitted = given - allowed
      raise ArgumentError, "Unsupported option(s): #{non_permitted.join(', ')}" if non_permitted.any?
    end

    def tokenize(text, options)
      text = pre_normalize(text)
      tokens = Tokenizer::tokenize(text)
      [Repeater, Grabber, Pointer, Scalar, Ordinal, Separator, Sign, TimeZone].each do |tok|
        tok.scan(tokens, options)
      end
      tokens.select { |token| token.tagged? }
    end

    def tokens_to_span(tokens, options)
      definitions = definitions(options)

      (definitions[:endian] + definitions[:date]).each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return handler.invoke(:date, good_tokens, self, options)
        end
      end

      definitions[:anchor].each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.select { |o| !o.get_tag Separator }
          return handler.invoke(:anchor, good_tokens, self, options)
        end
      end

      definitions[:arrow].each do |handler|
        if handler.match(tokens, definitions)
          good_tokens = tokens.reject { |o| o.get_tag(SeparatorAt) || o.get_tag(SeparatorSlash) || o.get_tag(SeparatorDash) || o.get_tag(SeparatorComma) || o.get_tag(SeparatorAnd) }
          return handler.invoke(:arrow, good_tokens, self, options)
        end
      end

      definitions[:narrow].each do |handler|
        if handler.match(tokens, definitions)
          return handler.invoke(:narrow, tokens, self, options)
        end
      end

      puts '-none' if Chronic.debug
      return nil
    end
  end
end
